File: felicia/examples/learn/topic/protobuf/cc/protobuf_subscribing_node.cc
#include "felicia/examples/learn/topic/protobuf/cc/protobuf_subscribing_node.h"

#include "felicia/core/master/master_proxy.h"

namespace felicia {

ProtobufSubscribingNode::ProtobufSubscribingNode(
    const TopicCreateFlag& topic_create_flag)
    : topic_create_flag_(topic_create_flag),
      topic_(topic_create_flag_.topic_flag()->value()) {}

void ProtobufSubscribingNode::OnInit() {
  std::cout << "ProtobufSubscribingNode::OnInit()" << std::endl;
}

void ProtobufSubscribingNode::OnDidCreate(const NodeInfo& node_info) {
  std::cout << "ProtobufSubscribingNode::OnDidCreate()" << std::endl;
  node_info_ = node_info;
  RequestSubscribe();

  // MasterProxy& master_proxy = MasterProxy::GetInstance();
  // master_proxy.PostDelayedTask(
  //     FROM_HERE,
  //     base::BindOnce(&ProtobufSubscribingNode::RequestUnsubscribe,
  //                      base::Unretained(this)),
  //     base::TimeDelta::FromSeconds(10));
}

void ProtobufSubscribingNode::OnError(const Status& s) {
  std::cout << "ProtobufSubscribingNode::OnError()" << std::endl;
  LOG(ERROR) << s;
}

void ProtobufSubscribingNode::OnMessage(SimpleMessage&& message) {
  std::cout << "ProtobufSubscribingNode::OnMessage()" << std::endl;
  std::cout << "message : " << message.DebugString() << std::endl;
}

void ProtobufSubscribingNode::OnMessageError(const Status& s) {
  std::cout << "ProtobufSubscribingNode::OnMessageError()" << std::endl;
  LOG(ERROR) << s;
}

void ProtobufSubscribingNode::OnRequestSubscribe(const Status& s) {
  std::cout << "ProtobufSubscribingNode::OnRequestSubscribe()" << std::endl;
  LOG_IF(ERROR, !s.ok()) << s;
}

void ProtobufSubscribingNode::OnRequestUnsubscribe(const Status& s) {
  std::cout << "ProtobufSubscribingNode::OnRequestUnsubscribe()" << std::endl;
  LOG_IF(ERROR, !s.ok()) << s;
}

void ProtobufSubscribingNode::RequestSubscribe() {
  communication::Settings settings;
  settings.buffer_size = Bytes::FromBytes(512);
  settings.channel_settings.tcp_settings.use_ssl =
      topic_create_flag_.use_ssl_flag()->value();

  subscriber_.RequestSubscribe(
      node_info_, topic_,
      ChannelDef::CHANNEL_TYPE_TCP | ChannelDef::CHANNEL_TYPE_UDP |
          ChannelDef::CHANNEL_TYPE_UDS | ChannelDef::CHANNEL_TYPE_SHM,
      settings,
      base::BindRepeating(&ProtobufSubscribingNode::OnMessage,
                          base::Unretained(this)),
      base::BindRepeating(&ProtobufSubscribingNode::OnMessageError,
                          base::Unretained(this)),
      base::BindOnce(&ProtobufSubscribingNode::OnRequestSubscribe,
                     base::Unretained(this)));
}

void ProtobufSubscribingNode::RequestUnsubscribe() {
  subscriber_.RequestUnsubscribe(
      node_info_, topic_,
      base::BindOnce(&ProtobufSubscribingNode::OnRequestUnsubscribe,
                     base::Unretained(this)));
}

}  // namespace felicia

File: felicia/examples/learn/topic/protobuf/python/protobuf_subscribing_node.py
import felicia_py as fel
from felicia.core.protobuf.channel_pb2 import ChannelDef

from felicia.examples.learn.topic.protobuf.simple_message_pb2 import SimpleMessage


class ProtobufSubscribingNode(fel.NodeLifecycle):
    def __init__(self, topic_create_flag):
        super().__init__()
        self.topic = topic_create_flag.topic_flag.value
        self.subscriber = fel.communication.Subscriber(SimpleMessage)

    def on_init(self):
        print("ProtobufSubscribingNode.on_init()")

    def on_did_create(self, node_info):
        print("ProtobufSubscribingNode.on_did_create()")
        self.node_info = node_info
        self.request_subscribe()

    def on_error(self, status):
        print("ProtobufSubscribingNode.on_error()")
        fel.log(fel.ERROR, status.error_message())

    def on_message(self, message):
        print("ProtobufSubscribingNode.on_message()")
        print("message : {}".format(message))

    def on_message_error(self, status):
        print("ProtobufSubscribingNode.on_message_error()")
        fel.log_if(fel.ERROR, not status.ok(), status.error_message())

    def on_request_subscribe(self, status):
        print("ProtobufSubscribingNode.on_request_subscribe()")
        fel.log_if(fel.ERROR, not status.ok(), status.error_message())

    def on_request_unsubscribe(self, status):
        print("ProtobufSubscribingNode.on_request_unsubscribe()")
        fel.log_if(fel.ERROR, not status.ok(), status.error_message())

    def request_subscribe(self):
        settings = fel.communication.Settings()
        settings.buffer_size = fel.Bytes.from_bytes(512)

        self.subscriber.request_subscribe(self.node_info, self.topic,
                                          ChannelDef.CHANNEL_TYPE_TCP |
                                          ChannelDef.CHANNEL_TYPE_UDP |
                                          ChannelDef.CHANNEL_TYPE_UDS |
                                          ChannelDef.CHANNEL_TYPE_SHM,
                                          settings,
                                          self.on_message, self.on_message_error,
                                          self.on_request_subscribe)

        # fel.MasterProxy.post_delayed_task(
        #         self.request_unsubscribe, fel.from_seconds(10))

    def request_unsubscribe(self):
        self.subscriber.request_unsubscribe(
            self.node_info, self.topic, self.on_request_unsubscribe)
